AttitudeManager/ControlAlgorithms/Src/AM_FixedControl.cpp

```diff
@@ -1,17 +1,116 @@
+    float groundSpeed;                  // m/s
+    double heading;                     // Degrees. Heading of 0 is north.
+} SFOutput_t;
 
 namespace AM {
 
 void FixedControl::runControlsAlgo(const AttitudeManagerInput &instructions,
                                    float outputs[],
                                    uint8_t outputsLength) const {
+   
+   // Get current attitude from sensorfusion 
+    SFOutput_t currentAttitude; // TODO: This needs to be retrieved from LOS
+    
+    // Compute target values
+    float targetHeading = currentAttitude.heading - instructions.heading;
+    float targetAltitude = instructions.z_dir; 
+    float targetPitch = instructions.x_dir * maxPitchAngle;
+    float targetBank = instructions.y_dir * maxBankAngle;
+     
+    // Adjust heading difference for the PID
+    if (targetHeading <= (-180)) {
+        targetHeading += 180;
+    } else if (targetHeading > 180){
+        targetHeading -=360;
+    }
+   
+    // Define PID values for mixing
+    PIDController pid_bank{bank_kp,        bank_ki,      bank_kd, 
+                           bank_i_windup, -maxBankAngle, maxBankAngle };
+    PIDController pid_rudder{rudder_kp,       rudder_ki, rudder_kd, 
+                             rudder_i_windup, -100,      100};
+    PIDController pid_pitch{pitch_kp,       pitch_ki,       pitch_kd, 
+                            pitch_i_windup, -maxPitchAngle, maxPitchAngle};
+    PIDController pid_airspeed{airspeed_kp,       airspeed_ki,  airspeed_kd, 
+                               airspeed_i_windup, airspeed_min, airspeed_max};                        
+
+    // Compute values for mixing
+    float bank = pid_bank.execute(targetBank, currentAttitude.roll, currentAttitude.rollRate);
+
+    float pitch = pid_pitch.execute(targetPitch, currentAttitude.pitch, currentAttitude.pitchRate);
+
+    float rudderSet = rudderPercent(bank);
+    
+    rudderSet = pid_rudder.execute(0.0f, targetHeading);
 
-    // do something
+    float airspeed = pid_airspeed.execute(targetAltitude, currentAttitude.altitude, // handling altitude with airspeed? Define altitude and airspeed seperately?
+                                          currentAttitude.rateOfClimb);
+    
+    float roll = DEG_TO_RAD(bank);
+
+    // mix the PID's
+    float engineOutput = 
+        mixPIDs(configs[Engine].stateMix, roll, pitch, rudderSet, airspeed); 
+    float leftAileronOutput = 
+        mixPIDs(configs[LeftAileron].stateMix, roll, pitch, rudderSet, airspeed);
+    float rightAileronOutput = 
+        mixPIDs(configs[RightAileron].stateMix, roll, pitch, rudderSet, airspeed);
+    float rudder = 
+        mixPIDs(configs[Rudder].stateMix, roll, pitch, rudderSet, airspeed);
+    float elevator = 
+        mixPIDs(configs[Elevator].stateMix, roll, pitch, rudderSet, airspeed);
+
+    // Verify values
+    assert(configs[Engine].channel < outputsLength); // Question: Engine controlled as some percentage? Indicative of airspeed?
```

Comment:
ok....how do you know how many configs there are...

Because we could have 2 right ailerons and 2 left ailerons and 2 rudders etc etc....

Have we thought of a way to use the statemix to assign these ? 

Same with mixPID's....we can iterate through each element in the config and use the statemix to assign output values.

IF you're not sure on how the statemix works, lmk and we can discuss more.

Reply:
More information on how the statemix works would be great. I think we could discuss this in our meeting tomorrow.

Reply:
A case of multiple pairs of ailerons would be a good case for a different control algorithm, right?